### PR TITLE
[FE] Feature: Features of the ward and district officer module

### PR DESCRIPTION
**Pull request by mitidevus**
### Task
Features of the ward officer

### Features
1) Locations
- View list
- View details
2) Advertising panels
- View list
- View details
- Create the DRAFT panel
- Send the licensing request for panel to CDO
3) Editing Request
- Send a request for editing the location to the CDO
- Send a request for editing the panel to the CDO
4) Report
- View list of location reports
- View list of location reports
- View details of location reports
- View details of panel reports
- Update status and solution for report
5) Requests
- View the list of licensing requests
- View details of licensing requests
- View the list of editing requests (Location)
- View details of editing request (Location)
- View the list of editing requests (Panel)
- View details of editing request (Panel)
- Cancel request (For unapproved requests)

### Screenshots
1) Locations
- View list
image

- View details
image

2) Advertising panels
- View list
image

- View details
image

- Create the DRAFT panel
image

- Send the licensing request for panel to CDO
image

3) Editing Request
- Send a request for editing the location to the CDO
image

- Send a request for editing the panel to the CDO
image

4) Report
- View list of location reports
image

- View list of location reports
image

- View details of location reports
image

- View details of panel reports
image

- Update status and solution for report
image

5) Requests
- View the list of licensing requests
image

- View details of licensing requests
image

- View the list of editing requests (Location)
image

- View details of editing request (Location)
image

- View the list of editing requests (Panel)
image

- View details of editing request (Panel)
image

- Cancel request (For unapproved requests)
image
image

### CODE REVIEW

File: apps/management-app/src/utils/format-date.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  const day = dateObj.getDate();
+  const month = dateObj.getMonth() + 1;
+  const year = dateObj.getFullYear();
+  const hours = dateObj.getHours();
+  const minutes = dateObj.getMinutes();
+  const seconds = dateObj.getSeconds();
+
+  return `${day}/${month}/${year} ${hours}:${minutes}:${seconds}`;
+}
```
Comment (starhuy1903) [inline apps/management-app/src/utils/format-date.ts:13]: try to use `moment` library

File: apps/management-app/src/components/Authenticated/AdsLocation/index.tsx [hunk 2]
```diff
@@ -1,5 +1,152 @@
...
+              <TableCell align="center">ID</TableCell>
+              <TableCell align="center">Address</TableCell>
+              <TableCell align="center">Ward</TableCell>
+              <TableCell align="center">District</TableCell>
+              <TableCell align="center">Position type</TableCell>
+              <TableCell align="center">Advertising type</TableCell>
+              <TableCell align="center">Planned</TableCell>
+              <TableCell align="center">Created</TableCell>
+              <TableCell align="center">Modified</TableCell>
+              <TableCell align="center"></TableCell>
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/AdsLocation/index.tsx:113]: create an array and map

File: apps/management-app/src/components/Authenticated/AdsPermission/index.tsx [hunk 3]
```diff
@@ -1,5 +1,263 @@
...
+      },
+      reason: 'Advertise products and services for the company',
+      status: 'Approved',
+      createdTime: '2023-12-08T11:30:53.945Z',
+      modifiedTime: '2023-12-08T11:30:53.945Z',
+    },
+  ];
+
+  const handleDelete = (id: number) => {
+    const confirm = window.confirm('Are you sure to delete this request?');
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/AdsPermission/index.tsx:182]: create Confirm modal

File: apps/management-app/src/components/Common/FormComponents.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+export function ReadOnlyTextForm({
+  label,
+  field,
+  value,
+}: {
+  label?: string;
+  field: string;
+  value?: string | number | undefined;
```
Comment (starhuy1903) [inline apps/management-app/src/components/Common/FormComponents.tsx:37]:
```suggestion
  value?: string | number;
```

File: apps/management-app/src/components/Common/FormComponents.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  TextField,
+} from '@mui/material';
+
+export function ReadOnlyTextField({
+  label,
+  value,
+  disabled = false,
+}: {
+  label?: string;
+  value?: string | number | undefined;
```
Comment (starhuy1903) [inline apps/management-app/src/components/Common/FormComponents.tsx:14]:
```suggestion
  value?: string | number;
```

File: apps/management-app/src/components/Common/FormComponents.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        readOnly: true,
+      }}
+      value={value}
+      disabled={disabled}
+    />
+  );
+}
+
+export function ReadOnlyTextForm({
+  label,
```
Comment (starhuy1903) [inline apps/management-app/src/components/Common/FormComponents.tsx:31]:
```suggestion
  label = '',
```
should add default value when it's optional field

File: apps/management-app/src/configurations/dev.ts
Comment (starhuy1903) [inline apps/management-app/src/configurations/dev.ts:1]: revert this file, run mock mode instead

File: apps/management-app/src/types/form.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  id: number;
+  panelType: string;
+  location: {
+    id?: number;
+    address: string;
+    ward: string;
+    commue: string;
+    positionType: string;
+    adsType: string;
+  };
```
Comment (starhuy1903) [inline apps/management-app/src/types/form.ts:26]: should create LocationType

File: apps/management-app/src/types/form.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  width: number;
+  height: number;
+  quantity: number;
+  imageUrl: string;
+  company: {
+    email: string;
+    phone: string;
+    createdContractDate: string;
+    expiredContractDate: string;
+  };
```
Comment (starhuy1903) [inline apps/management-app/src/types/form.ts:36]: create CompanyInfo type

File: apps/management-app/src/components/Authenticated/Officer/LocationReportList/index.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+export default function LocationReportList() {
+  const [page, setPage] = useState<number>(1);
+  const [reports, setReports] = useState<Report[] | undefined>([]);
+
+  const { data, isLoading, refetch } = useGetReportsQuery({
+    page: page,
+    take: 10,
+    targetType: 'Location',
+    typeId: 1,
+  });
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/Officer/LocationReportList/index.tsx:41]: useLazy request

File: apps/management-app/src/components/Authenticated/Officer/PanelCreating/index.tsx [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+      ).toISOString();
+      formData.append('createContractDate', data.createContractDate);
+      formData.append('expiredContractDate', data.expiredContractDate);
+      data.images.forEach((image) => formData.append('images', image));
+
+      await createPanel(formData).unwrap();
+
+      setSubmitting(false);
+
+      navigate(-1);
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/Officer/PanelCreating/index.tsx:138]: should redirect to specific route

File: apps/management-app/src/components/Authenticated/Officer/PanelDetail/index.tsx [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  const [panel, setPanel] = useState<Panel | undefined>(undefined);
+  const { panelId } = useParams<{ panelId: string }>();
+  const { data, isLoading, refetch } = useGetPanelByIdQuery(panelId!);
+
+  useEffect(() => {
+    refetch();
+  }, [panel, refetch]);
+
+  useEffect(() => {
+    if (data) {
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/Officer/PanelDetail/index.tsx:25]: useLazy request

File: apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+  const navigate = useNavigate();
+
+  const { panelId } = useParams<{ panelId: string }>();
+
+  const userId = useAppSelector((state) => state?.user?.profile?.id);
+
+  const [panel, setPanel] = useState<Panel | undefined>(undefined);
+  const [panelTypes, setPanelTypes] = useState<PanelType[]>([]);
+
+  const { data: panelData, isLoading: panelLoading } = useGetPanelByIdQuery(
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx:57]: useLazy request

File: apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+      formData.append('createContractDate', data.createContractDate);
+      formData.append('expiredContractDate', data.expiredContractDate);
+      formData.append('companyEmail', data.companyEmail);
+      formData.append('companyNumber', data.companyNumber);
+
+      await updatePanel(formData).unwrap();
+
+      setSubmitting(false);
+
+      navigate(-1);
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx:174]: refactor this and add showSuccess toaster on updated API

**Review by starhuy1903 — CHANGES_REQUESTED** [review]

File: apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx (hunk 13 above)
Comment (mitidevus) [reply apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx:174]: Can we use **getOnMutationFunction** in _helper.ts_?
image
image

Comment (starhuy1903) [reply apps/management-app/src/components/Authenticated/Officer/PanelEditing/index.tsx:174]: should `showSuccess` in component and we will refactor all code later

**Review by starhuy1903 — APPROVED** [review]

**Pull request merged**